Move lib and scan modules into a package

thewilsonator opened the pull request:
The only symbols declared in these modules used outside them are declared in `src/lib.d`.

This commit changes:
* `lib{elf,mach,mscoff,omf}.d` => `lib/{elf,mach,mscoff,omf}.d`: Those files are only imported by a single module, dmd.lib;
* `lib.d` => `lib/package.d`: The single module importing `lib*.d`, which provides the single entry point for this package;
`scan*.d` => `lib/scan*.d`: Those files are never imported by modules except for the corresponding `lib/{elf,mach,mscoff,omf}.d`;

This increases encapsulation by:
* reducing the visibility of symbols in `scan*.d` to `dmd.lib`, so the context one needs looking at the module in isolation is reduced;
* it reduces the visibility of symbols in `{elf,mach,...}.d`;
* It signals to the reader that `package.d` is the entry point, based on a language feature, not coding convention

I hope I have all the doc links correct.

WalterBright commented:
We've extensively discussed this. Not doing packages.

thewilsonator commented:
Nope

wilzbach reviewed (approved):
There's zero harm in them and it drastically helps newcomers to get around in the code base.

12345swordy commented:
> 
> 
> We've extensively discussed this. Not doing packages.

Walter, as a newcomer to the dmd codebase, I legitimately don't understand why you are so adamant of being against organizing the files in a way that promotes encapsulation, by using packages here. I have talk about it in discord where the expresstionsem file seemed to be importing modules from everywhere with no regards to being organized.

WalterBright commented:
We've been over that many times, including yet another discussion about it recently in the n.g.

WalterBright reviewed (dismissed):
Not approved. See many discussions about this, in the n.g. and at DConf.

ibuclaw commented on src/dmd/lib/elf.d:
```diff
@@ -38,10 +38,10 @@ import dmd.root.rmem;
 import dmd.root.string;
 import dmd.root.stringtable;
```
From memory, this is one of the big reasons why this isn't a package.

Geod24 replied:
I'm not sure I follow your comment here?

ibuclaw replied:
dmd.root is a logical package, and doesn't depend on anything outside of it save for druntime bindings (i.e another project) - the same goes for dmd.backend as well (this is why there are two OutBuffer types, one in root and the other in backend).

By contrast dmd.lib as a package would import both from its parent dmd package and sibling dmd.root package (so, no encapsulation at all).

thewilsonator replied:
> By contrast dmd.lib as a package would import both from its parent dmd package and sibling dmd.root package 

So what? 

> (so, no encapsulation at all).

What? That makes no sense. 

Packaging is about:
* grouping modules logically (lib, visitors, AST etc.)
* symbols internal to the package not being exposed (`package` level visibility)
* package level dependence (and more importantly lack thereof, e.g. AST should not depend on semantic analysis).

Geod24 replied:
@ibuclaw : It sounds like you are advocating for no cross-package import ? If so, that's pretty unworkable (and IMO not how package were intended to be used).

ibuclaw replied:
> > (so, no encapsulation at all).
> 
> What? That makes no sense.
> 
> Packaging is about:
> 

This has been repeated ad-nauseam, so might as well just [cross reference](https://forum.dlang.org/post/s8ftnv$aoq$1@digitalmars.com) at this point.

ibuclaw replied:
> @ibuclaw : It sounds like you are advocating for no cross-package import ? If so, that's pretty unworkable (and IMO not how package were intended to be used).

It is workable, and [it can be done](https://github.com/dlang/dmd/pull/12537) without imports.

ibuclaw replied:
To be clear, I'm totally for making packages, but I also appreciate the rationale against doing it like this PR.  Agree with it or not, Walter does have a good point here, and you are just going to have to adapt and come up with something that overcomes that disagreement, rather than keep being dismissive about it.

Geod24 replied:
For reference, the quote from @WalterBright is:
> To be a little clearer, if the files are all merely reshuffled into various packages, then they all violate the rule:
>
>    *** Never import a file from an uplevel directory ***
>
> and understanding is not increased at all. **And it isn't even just an uplevel directory, it's up then sideways then down.**

Emphasis mine.

This is the silliest rule ever. And it doesn't make sense. Absolutely no code in D (that uses packages) respects it. Look at Phobos, druntime, whatever project, they import sideways. Or is it okay if the top level package is not shared ? This rule is a self-fulfilling restriction for flat package hierarchy, because importing *down* is *completely backwards*.

Packages are trees. The root of the tree is more generic than the leaves, e.g. `std` is more generic than `std.algorithm` which is more generic than `std.algorithm.sorting`.  The abstraction should be at a higher package level, the implementation at a lower level, hence this rule can be essentially rephrased as "the implementation should not import the abstraction".

Geod24 replied:
@ibuclaw : Remember https://github.com/sociomantic-tsunami/ocean/issues/573 ?
@donc points to the same issue in the first bullet. Having the abstraction at a lower level than the implementation is nonsense.

dkorpel replied:
It's also worth mentioning that dmd/root "was meant as a portable library used by many projects, but it wasn't very good and the only project left using it is dmd". It makes no sense that the lib package is allowed to import `core.stdc.stdlib: strtold` because it's a 'separate project', but not `dmd.root.strtold` because it happens to be in the dmd namespace and violates a "no sideways imports" rule.

MoonlightSentinel replied:
> It is workable, and [it can be done](https://github.com/dlang/dmd/pull/12537) without imports.

Templatizing something only to import declarations through an intermediate type is still an import, it's just wrapped nicely in gift paper. The dependencies are still there, just hidden by an additional layer.

thewilsonator replied:
Resolving this

Geod24 commented on src/dmd/lib/scanomf.d:
```diff
@@ -22,6 +22,8 @@ import dmd.errors;
 
 private enum LOG = false;
 
+package(dmd.lib):
```
This can just be `package`

thewilsonator replied:
Done, applied throughout.

Geod24 commented:
> We've been over that many times, including yet another discussion about it recently in the n.g.

Well you have been known to change your mind from time to time, I'm sure @thewilsonator had hope this was one of those cases :)

The discussion on the n.g. boiled down to encapsulation. My takeaway from it - correct me if I'm wrong - was that packaging should be done to increase encapsulation, and that spilling modules into package that imports each other is a waste of time.

But what @thewilsonator chose as a starting point is a textbook example of packages.
The file changes are:
- `lib{elf,mach,mscoff,omf}.d` => `lib/{elf,mach,mscoff,omf}.d`: Those files are only imported by a single module, `dmd.lib`;
- `lib.d` => `lib/package.d` => The single module importing `lib*.d`, which provides the single entry point for this package;
- `scan*.d` => Those files are never imported by modules except for the corresponding `lib{elf,mach,mscoff,omf}.d`;

In that regard, making `lib` a package increases encapsulation, because:
- It reduces the visibility of symbols in `scan*.d` to `dmd.lib`, so one looking at the module in isolation needs a reduced context;
- Likewise, it reduces the visibility of symbols in `{elf,mach,...}.d`;
- It signal to the reader that `package.d` is the entry point, based on a language feature, not coding convention;

I'm still very doubtful of some of the packages structures that were floated around in the newsgroup, but this is an obvious case where packages can be used.

thewilsonator commented:
> I'm sure @thewilsonator had hope this was one of those cases :)

On the contrary https://github.com/dlang/dmd/pull/12744#issuecomment-868901268 the number of files in `src/dmd` is getting even sillier.

> The discussion on the n.g. boiled down to encapsulation. My takeaway from it - correct me if I'm wrong - was that packaging should be done to increase encapsulation, and that spilling modules into package that imports each other is a waste of time.

Indeed, there are only two sets of modules that are close to being as encapsulated as they can possibly be and fit for becoming packages at the moment:
* `lib*`/`scan*` which depend on `dmd.root`, `dmd.utils` (for path stuff) and `dmd.{global, errors}` for error reporting
* the various visitor modules, which depend on the AST only

PetarKirov commented:
@thewilsonator can you incorporate @Geod24's explanation of the changes in the commit message and PR description? Otherwise, obviously LGTM.

WalterBright commented:
> Absolutely no code in D (that uses packages) respects it.

That's mostly right. I've attempted to prevent a lot of it, such as repeated attempts to have the backend code import from the front end, and root importing from the dmd level.

dmd level modules each import every other dmd module (more or less). There is no hierarchy. At one time cparse.d imported target.d, which transitively imports everything else in dmd. @ibuclaw and I worked together to successfully remove that dependency. We could go further and eliminate nearly all dependencies from target.d. There's no reason target.d should be importing statement.d.

*That* sort of work will be worth while. Not having a fake package hierarchy that the code implementation does not remotely respect.

For people who want an overview of what the files are, see https://github.com/dlang/dmd/blob/master/src/dmd/README.md which does a more readable and thorough job than any misleading, fake package hierarchy. It could be improved with links to the generated Ddoc documentation files for each module. (Which is why I harangue PR submitters to add proper Ddoc function comments.)

I've said all this before, repeatedly.

BTW, something that *would* make dmd much easier to work on would be:

1. removing the obvious cut/paste code that was put into astbase.d. I did some of it already with astenums.d, which saved us from hundreds of lines of stinky copy/pasta code.
2. auto-generate the .h files, like symbol.h, instead of doing them manually. Inconsistencies have been causing bugs constantly for 10-15 years now. @ibuclaw just submitted another fix for one in scope.h. This isn't just a code smell, it's a code sewer.

Geod24 commented:
> I've attempted to prevent a lot of it, such as repeated attempts to have the backend code import from the front end, and root importing from the dmd level.

**Something we can all agree on.** I can't recall a seasoned DMD contributors arguing against this rule. I do recall having broken this rule myself, by mistake. This was corrected by @ibuclaw in https://github.com/dlang/dmd/pull/10720 and as you can see, there was no disagreement that it was the correct course of action.

But that does not in any way justify "Never import a file from an uplevel directory". It doesn't refute any of the arguments that were made, by me or other contributors.

> dmd level modules each import every other dmd module (more or less). There is no hierarchy. [...]

There is no *apparent* hierarchy to the modules. But there is an *inherent* hierarchy to DMD's code, otherwise it would be an unworkable mess. There are obvious separation between different components of DMD: the driver (`mars.d` & affiliates), frontend, IR, backend. There is also an inherent separation within those components: the AST class hierarchy is the most obvious example, but we also have a hierarchy of visitors.

What us contributors are asking for is to be able to express this separation using D's native unit of encapsulation, a.k.a. the module. (While I'm using the term "module" here, it encompass packages as well - Modules can only scale when used with packages).

> Not having a fake package hierarchy that the code implementation does not remotely respect.

What is fake about `dmd.lib` ? I gave a thorough explanation about why it can be a package [here](https://github.com/dlang/dmd/pull/12766#issuecomment-869242653).

> For people who want an overview of what the files are [..]

READMEs and conventions are no substitutes for using languages features for their intended use.
To make an analogy, why do we need the `deprecated` attribute if we can just mark functions as deprecated in the documentation ? And even if we agree that `deprecated` is needed, why do we need to add a message to it, users can just look up the documentation when they see the message ?
Documentation is a supplement, not a replacement.

> removing the obvious cut/paste code that was put into astbase.d. I did some of it already with astenums.d, which saved us from hundreds of lines of stinky copy/pasta code.

That's good. And can be pursued in parallel of packaging. The two do not conflict, nor is one a substitute for the other.

> auto-generate the .h files, like symbol.h, instead of doing them manually.

This has been worked on by contributors for years now, under the `dtoh` name.
This led to the dreaded `frontend.h`which you [wanted to remove](https://github.com/dlang/dmd/pull/11910). However we can't fully switch to using it until it is completely stable and usable, and auto-generating `.h` files that can accurately represent all the constructs that are used in a codebase as large and diverse as DMD is no small task.
For ref, just over the past 2 weeks, there have been 9 PRs raised or merged in support of this.

WalterBright commented:
> This led to the dreaded frontend.h which you wanted to remove.

This is a misunderstanding. I wanted to remove it from the git source code base, and have it be auto-generated as part of the build process instead.

> dtoh

I'm very glad to see progress on this front. All the .h files need to be removed and auto-generated. It's the flip side of ImportC, which will hopefully eliminate most of the .h => .d conversion nuisance and consequent errors. We can't get to 100% with ImportC due to endless C extensions and things in C that are not representable in D. But we *should* be able to do 100% for .h file emission for `extern(C)` and `extern(C++)` functions.

ibuclaw commented:
> > This led to the dreaded frontend.h which you wanted to remove.
> 
> This is a misunderstanding. I wanted to remove it from the git source code base, and have it be auto-generated as part of the build process instead.
> 
> > dtoh
> 
> I'm very glad to see progress on this front. All the .h files need to be removed and auto-generated.

I still have occasional doubts about whether integrating it into the compiler was the right choice. I originally had it as a dedicated external tool aimed specifically at translating dmd codebase by having inside knowledge about what  equivalent special types are used internally, etc.  By the time I stopped working on it, it was able to produce something that was 100% valid and compiled with the test C++ front-end.

Whilst generalising it has no doubt made it better for all, it has regressed somewhat on its ability to translate dmd.  I do fear that the choice to move it inside the compiler is holding it back as well, though. For instance, why translate a non-trivially templated type (and all its forward referenced dependencies) when a `void*` will do just fine for its intended purpose? If only we were able to give it hints about what to generate via a crafted code comment format (`/// dtoh: void*`) or maybe something like a map file to tell it what to do when it encounters something like `Identifier[Array!const(char)[]]*`.

Alas, I digress. :-)

MoonlightSentinel commented:
> For instance, why translate a non-trivially templated type (and all its forward referenced dependencies) when a `void*` will do just fine for its intended purpose?

`dtoh` already uses forward declarations whenever possible to skip/defer the actual declaration. It will only emit `TemplateDeclaration`'s if either an instance is used by value (e.g. void foo(`Array!int)` ) or the template is explicitly marked as `extern(C++)`. The latter could even be ignored because `dtoh` can resolve the dependendency on first use.

> Alas, I digress. :-)

Indeed. The whole discussion about `astbase.d` and header files is moot because they imply some maintenance that is either automated or done by copy + pasting some declarations. This can be a nuisance but implies that you already made significant changes (worthy of a PR) and hence already did >95% of the work. 

This PR improves the structure of the code base to make the 95% easier for contributors who have not yet been working on DMD for decades.

MoonlightSentinel reviewed (approved):
Rationale outlined in https://github.com/dlang/dmd/pull/12766#issuecomment-869242653

thewilsonator committed:
Move lib and scan modules into a package

The only symbols declared in these modules used outside them are declared in `src/lib.d`.

This commit changes:
* `lib{elf,mach,mscoff,omf}.d` => `lib/{elf,mach,mscoff,omf}.d`: Those files are only imported by a single module, dmd.lib;
* `lib.d` => `lib/package.d`: The single module importing `lib*.d`, which provides the single entry point for this package;
`scan*.d` => `lib/scan*.d`: Those files are never imported by modules except for the corresponding `lib/{elf,mach,mscoff,omf}.d`;

This increases encapsulation by:
* reducing the visibility of symbols in `scan*.d` to `dmd.lib`, so the context one needs looking at the module in isolation is reduced;
* it reduces the visibility of symbols in `{elf,mach,...}.d`;
* It signals to the reader that `package.d` is the entry point, based on a language feature, not coding convention;

thewilsonator commented:
> @thewilsonator can you incorporate @Geod24's explanation of the changes in the commit message and PR description? Otherwise, obviously LGTM.

Done

PetarKirov commented:
@WalterBright I think it's about time we embrace the D way of doing things in dmd :P

We wouldn't be in the worst company:

Clang C++ compiler: https://github.com/llvm/llvm-project/tree/llvmorg-12.0.0/clang/lib
C# compiler: https://github.com/dotnet/roslyn/tree/main/src/Compilers/CSharp/Portable
Dart compiler: https://github.com/dart-lang/sdk/tree/2.13.0/pkg
Rust compiler: https://rustc-dev-guide.rust-lang.org/compiler-src.html#compiler https://github.com/rust-lang/rust/tree/master/compiler
Swift compiler: https://github.com/apple/swift/tree/main/lib

WalterBright commented:
> But that does not in any way justify "Never import a file from an uplevel directory".

Good question. What dmd has is the package hierarchy literally upside down. May I suggest the chapter "The Dependency-Inversion Principle" in Robert Martin's book "Agile Software Development."

https://www.amazon.com/Software-Development-Principles-Patterns-Practices/dp/0135974445

> This PR improves the structure of the code base to make the 95% easier for contributors who have not yet been working on DMD for decades.

https://github.com/dlang/dmd/blob/master/src/dmd/README.md

Geod24 commented on src/dmd/lib/package.d:
```diff
@@ -1,13 +1,13 @@
 /**
  * A module defining an abstract library.
- * Implementations for various formats are in separate `libXXX.d` modules.
+ * Implementations for various formats are in `src/dmd/lib/` modules.
```
Well technically that's not quite accurate:
- There is a 1-to-1 mapping between modules and files, but they are not the same concept. When referring to modules, we use the name in the `ModuleDeclaration`, e.g. `dmd.lib.elf`;
- `dmd.lib` still contains `scan*` which are not implementation, more like helpers;

But since it's consistent with the previous wording, we can fix it later. I think we have an opportunity to remove the function-backed factory now that we are going with proper encapsulation.

Geod24 reviewed (approved):
LGTM

dkorpel commented on src/dmd/README.md:
```diff
@@ -183,15 +183,15 @@ Note that these groups have no strict meaning, the category assignments are a bi
-| [scanmach.d](https://github.com/dlang/dmd/blob/master/src/dmd/scanmach.d)     | Extract symbol names from a library in Mach-O format |
-| [scanmscoff.d](https://github.com/dlang/dmd/blob/master/src/dmd/scanmscoff.d) | Extract symbol names from a library in COFF format   |
-| [scanomf.d](https://github.com/dlang/dmd/blob/master/src/dmd/scanomf.d)       | Extract symbol names from a library in OMF format    |
+| [lib/package.d](https://github.com/dlang/dmd/blob/master/src/dmd/lib/package.d)               | Abstract library class                               |
+| [lib/elf.d](https://github.com/dlang/dmd/blob/master/src/dmd/lib/elf.d)         | Library in ELF format (Unix)                         |
+| [lib/mach.d](https://github.com/dlang/dmd/blob/master/src/dmd/lib/mach.d)       | Library in Mach-O format (macOS)                     |
+| [lib/mscoff.d](https://github.com/dlang/dmd/blob/master/src/dmd/lib/mscoff.d)   | Library in COFF format (32/64-bit Windows)           |
+| [lib/omf.d](https://github.com/dlang/dmd/blob/master/src/dmd/lib/omf.d)         | Library in OMF format (legacy 32-bit Windows)        |
+| [lib/scanelf.d](https://github.com/dlang/dmd/blob/master/src/dmd/lib/scanelf.d)       | Extract symbol names from a library in ELF format    |
+| [lib/scanmach.d](https://github.com/dlang/dmd/blob/master/src/dmd/lib/scanmach.d)     | Extract symbol names from a library in Mach-O format |
+| [lib/scanmscoff.d](https://github.com/dlang/dmd/blob/master/src/dmd/lib/scanmscoff.d) | Extract symbol names from a library in COFF format   |
+| [lib/scanomf.d](https://github.com/dlang/dmd/blob/master/src/dmd/lib/scanomf.d)       | Extract symbol names from a library in OMF format    |
```
I think you can remove this entire table and simply add the 'lib' folder to 'Directory structure' at the start of the README.

RazvanN7 commented:
@WalterBright Given that this PR has received 6 approvals from community members, are you inclined on accepting this?

nordlow commented:
@WalterBright what are the pros of not encapsulating these files into packages?
